features/keychain/module/keychain.js

```diff
@@ -44,30 +44,50 @@ export class Keychain {
     this.secp256k1 = secp256k1.create({ getPrivateHDKey: this.#getPrivateHDKey })
   }
 
-  #assertPrivateKeysUnlocked() {
-    assert(!this.#privateKeysAreLocked, 'private keys are locked')
+  #assertPrivateKeysUnlocked(seedIds) {
+    const locked = this.#checkPrivateKeysLocked(seedIds)
+    assert(!locked, 'private keys are locked')
   }
 
-  arePrivateKeysLocked() {
-    return this.#privateKeysAreLocked
+  #checkPrivateKeysLocked(seedIds) {
+    if (!seedIds?.size) {
+      return Object.values(this.#seedLockStatus).some(Boolean)
+    }
+
+    const existingSeeds = Object.entries(this.#seedLockStatus)
+    const checkingSeeds = existingSeeds.filter(([seedId]) => seedIds.has(seedId))
+
+    assert(checkingSeeds.length === seedIds.size, 'must pass in existing seeds')
+    return checkingSeeds.some(([, locked]) => locked)
+  }
+
+  arePrivateKeysLocked(seeds = []) {
+    const seedIds = getManySeedIds(seeds)
+    return this.#checkPrivateKeysLocked(seedIds)
   }
 
   lockPrivateKeys() {
-    this.#privateKeysAreLocked = true
+    const seedIds = Object.keys(this.#masters)
+    this.#seedLockStatus = seedIds.reduce((acc, seedId) => {
+      acc[seedId] = true
+      return acc
+    }, Object.create(null))
   }
 
-  unlockPrivateKeys(seeds) {
-    assert(this.#privateKeysAreLocked, 'already unlocked')
-    assert(
-      seeds?.length === Object.values(this.#masters).length,
-      'must pass in same number of seeds'
-    )
-    const seedIds = new Set(seeds.map((seed) => getSeedId(seed)))
-    for (const seedId of Object.keys(this.#masters)) {
-      assert(seedIds.has(seedId), 'must pass in existing seed')
+  unlockPrivateKeys(seeds = []) {
+    const seedIds = getManySeedIds(seeds)
+    const locked = this.#checkPrivateKeysLocked(seedIds)
+    assert(locked, 'already unlocked')
```

Comment:
I am wondering if we should allow this method to be idempotent and therefore consumers not having to track which keys are currently locked. I imagine consumers would be mainly interested in the outcome

```
- A: locked, B: unlocked, C: locked
- unlockPrivateKeys([A, B])
- A: unlocked, B: unlocked, C: locked
- unlockPrivateKeys([A, B])  -> does not throw
- A: unlocked, B: unlocked, C: locked

Reply:
I removed the check, it should now be possible to call `unlock` on a already unlocked seed